lib/layout.js

```diff
@@ -191,44 +402,200 @@ export default class PodiumLayout {
+     * * Mounts a proxy so that each podlet can do transparent proxy requests as needed.
+     * * Returns a Promise which will resolve with the HttpIncoming object that was passed in.
+     *
+     * If the inbound request matches a proxy endpoint the returned Promise will resolve with a HttpIncoming object where the .proxy property is set to true.
+     *
+     * @see https://podium-lib.io/docs/api/layout#processhttpincoming
+     * @see https://podium-lib.io/docs/api/incoming
+     *
+     * @example
+     * ```js
+     * import { HttpIncoming } from '@podium/utils';
+     * import Layout from '@podium/layout';
+     *
+     * const layout = new Layout({
+     *     name: 'myLayout',
+     *     pathname: '/',
+     * });
+     *
+     * const server = http.createServer(async (req, res) => {
+     *     const incoming = new HttpIncoming(req, res);
+     *
+     *     try {
+     *         const result = await layout.process(incoming);
+     *         if (result.proxy) return;
+     *
+     *         res.statusCode = 200;
+     *         res.setHeader('Content-Type', 'application/json');
+     *         res.end(JSON.stringify(result));
+     *     } catch (error) {
+     *         res.statusCode = 500;
+     *         res.setHeader('Content-Type', 'text/plain');
+     *         res.end('Internal server error');
+     *     }
+     * });
+     * ```
+     *
+     * @param {HttpIncoming} incoming
+     * @param {{ proxy?: boolean, context?: boolean }} [options]
+     * @returns {Promise<HttpIncoming>}
+     */
     async process(incoming, { proxy = true, context = true } = {}) {
         incoming.name = this.name;
         incoming.css = [...this.cssRoute];
         incoming.js = [...this.jsRoute];
 
+        // @ts-ignore
         if (context) await this.context.process(incoming);
         if (proxy) await this.httpProxy.process(incoming);
 
         return incoming;
     }
 
+    /**
+     * A Connect/Express compatible middleware function which takes care of the various operations needed for a layout to operate correctly. This function is more or less just a wrapper for the .process() method.
+     * The middleware will create an HttpIncoming object for each request and place it on the response at res.locals.podium.
+     *
+     * **Important:** *This middleware must be mounted before defining any routes.*
+     *
+     * @see https://podium-lib.io/docs/api/layout#middleware
+     *
+     * @example
+     * ```js
+     * const app = express();
+     * app.use(layout.middleware());
+     * ```
+     *
+     * @returns {(req: any, res: any, next: function) => Promise<void>}
+     */
     middleware() {
         return async (req, res, next) => {
-            const incoming = new HttpIncoming(req, res, res.locals);
-            incoming.url = `${req.protocol}://${req.get('host')}${req.originalUrl}`;
+            const incoming = new HttpIncoming(req, res);
+            // @ts-ignore
+            incoming.url = new URL(
```

Comment:
This changes the type from a `string` to `URL`. I'm a bit skeptical, this might break some things downstream.

Reply:
Good point!